File: JS/CountingSort.js
function countingSort(arr) {
  const n = arr.length;

  // Find the maximum element in the array to determine the range of values.
  let max = arr[0];
  for (let i = 1; i < n; i++) {
    if (arr[i] > max) {
      max = arr[i];
    }
  }

  // Create a count array to store the count of each element.
  const count = new Array(max + 1).fill(0);

  // Count the occurrences of each element.
  for (let i = 0; i < n; i++) {
    count[arr[i]]++;
  }

  // Reconstruct the sorted array from the count array.
  let outputIndex = 0;
  for (let i = 0; i <= max; i++) {
    while (count[i] > 0) {
      arr[outputIndex] = i;
      outputIndex++;
      count[i]--;
    }
  }
}

// Example usage:
const arr = [4, 2, 2, 8, 3, 3, 1];
countingSort(arr);
console.log("Sorted array: " + arr);

File: Dart/CountingSort.dart
void countingSort(List<int> arr) {
  int n = arr.length;
  int max = arr.reduce((curr, next) => curr > next ? curr : next);
  int min = arr.reduce((curr, next) => curr < next ? curr : next);
  int range = max - min + 1;

  List<int> count = List<int>.filled(range, 0);
  List<int> output = List<int>.filled(n, 0);

  for (int i = 0; i < n; i++) {
    count[arr[i] - min]++;
  }
  for (int i = 1; i < range; i++) {
    count[i] += count[i - 1];
  }
  for (int i = n - 1; i >= 0; i--) {
    output[count[arr[i] - min] - 1] = arr[i];
    count[arr[i] - min]--;
  }
  for (int i = 0; i < n; i++) {
    arr[i] = output[i];
  }
}
